Improve Intel hardware intrinsic APIs

fiigii opened the pull request:
This PR
1. fixes https://github.com/dotnet/coreclr/issues/17058 and temporarily disables AVX MaskLoad test cases.
2. <del>encodes the result flags in the names of certain SSE4.2 string processing intrinsic. That provides more stable runtime behaviors and simplifies JIT implementation, discussed in https://github.com/dotnet/coreclr/issues/16270</del>
3. Fixes SSE4.1 `Insert` API on `float` base type, closes https://github.com/dotnet/coreclr/issues/18143.
4. Fixes SSE2/SSE4.1/AVX `Extract` return type to reflect the underlying instruction behavior, closes https://github.com/dotnet/coreclr/issues/17957

@CarolEidt @tannergooding @eerhardt

4creators commented:
The new functions names are really horrible and in majority of cases do not follow .NET API coding guidelines by using acronyms.

eerhardt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse42.cs:
```diff
@@ -18,115 +18,242 @@ public static class Sse42
-        ///   PCMPISTRI xmm, xmm/m128, imm8
-        /// int _mm_cmpistrs (__m128i a, __m128i b, const int imm8)
+        /// </summary>
+        public static bool CompareImplicitLengthNotCAndNotZ(Vector128<sbyte> left, Vector128<sbyte> right, StringComparisonMode mode) => CompareImplicitLengthNotCAndNotZ(left, right, mode);
```
Do we really need the words `ImplicitLength` and `ExplicitLength` in the method names?  Isn't the fact that I used the overload without specifying the lengths enough to disambiguate that I wanted to use implicit lengths?

fiigii replied:
Good point! Will try to eliminate these two words.

eerhardt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse42.cs:
```diff
@@ -18,115 +18,242 @@ public static class Sse42
         ///   PCMPISTRI xmm, xmm/m128, imm8
-        /// int _mm_cmpistrz (__m128i a, __m128i b, const int imm8)
+        /// </summary>
+        public static bool CompareImplicitLengthNotCAndNotZ(Vector128<byte> left, Vector128<byte> right, StringComparisonMode mode) => CompareImplicitLengthNotCAndNotZ(left, right, mode);
```
What does `C` and `Z` mean?

fiigii replied:
> What does C and Z mean?
> What does O mean? Similiarly, what does the S suffix mean below?

They all represent the corresponding flags in the [FLAGS register](https://en.wikipedia.org/wiki/FLAGS_register) of x86/x64 CPUs.

>  don't use single letter abbreviations for prefixes/suffixes in names.

I agree, but now we have some similar APIs that "use single letter abbreviations", such as `Sse41.TestC`, `Avx.TestZ`, etc. 
Do you have any suggestion to improve the names?

eerhardt replied:
I assume `C` == `Carry`, `Z` == `Zero`, `O` == `Overflow`, and `S` == `Sign`?  Can we spell out the words?  Or even better, if we could come up a more descriptive name of what the method does, that would be great.

Reading https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&text=_mm_cmpestr&expand=814,813,814,813, the only difference between `S` and `Z` that I can tell is which argument to check for a null character:

```
Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in a was null, and 0 otherwise. 
```
vs.
```
Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise. 
```

It seems the name should indicate/describe the differences, instead of using `SFlag` or `ZFlag`, and then forcing the caller to look up what `SFlag` means.

eerhardt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse42.cs:
```diff
@@ -18,115 +18,242 @@ public static class Sse42
+        /// int _mm_cmpistro (__m128i a, __m128i b, const int imm8)
+        ///   PCMPISTRI xmm, xmm/m128, imm8
+        /// </summary>
+        public static bool CompareImplicitLengthO(Vector128<byte> left, Vector128<byte> right, StringComparisonMode mode) => CompareImplicitLengthO(left, right, mode);
```
What does `O` mean?  Similiarly, what does the `S` suffix mean below?  I agree with @4creators that we typically don't use single letter abbreviations for prefixes/suffixes in names.

eerhardt commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Avx2.cs:
```diff
@@ -544,243 +544,243 @@ public static class Avx2
         ///   VPGATHERDD xmm, vm32x, xmm
         /// </summary>
-        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
+        public static unsafe Vector128<int> GatherVector128WithVector128Int32Indices(int* baseAddress, Vector128<int> index, byte scale) => GatherVector128WithVector128Int32Indices(baseAddress, index, scale);
```
While I can imagine it being easier for JIT implementation, these names are a bit unwieldy from a user's point of view IMO.

fiigii replied:
Yes, but if we only rely on overload to distinguish these intrinsics, the current named-intrinsic framework of RyuJIT has to be changed. Meanwhile, in the future, other platforms (e.g., .NET Native, Mono, etc) are also hard to port hardware intrinsic.

JIT experts may have some thoughts? @CarolEidt @mikedn @AndyAyersMS

jkotas replied:
The JIT intrinsic plumbing needs to be able to deal with overloads. It does it today in several cases.  For example, `CORINFO_INTRINSIC_Abs` can be either `Abs(float)` or `Abs(double)`.  Or there are multiple intrinsic overloads of `Vector::.ctor`. If the named intrinsics are meant to be the one true way to identify intrinsics in future, it should be just fixed there.

> Meanwhile, in the future, other platforms (e.g., .NET Native, Mono, etc) are also hard to port hardware intrinsic.

I do not think so.

CarolEidt replied:
It doesn't seem to me that it will be all that difficult to distinguish these, and given that it is not a common case, we could either choose to simply generate different `NI*` intrinsic names in the importer, based on the type of the relevant argument, **or** we could postpone that analysis until codegen, and look at the base type of the SIMD argument to determine which instruction to generate.

CarolEidt replied:
And I would say that, in general, the API shape is more important than the complexity of the JIT implementation, as long as it is both "pay for play" (i.e. the complexity is only incurred for the case in question) and the complexity is not very high.

fiigii replied:
Thanks, will retain the current APIs and address in JIT.

fiigii commented:
Re-improved SSE4.1 intrinsic APIs (e.g., `CompareImplicitLengthC` -> `CompareCFlag`) based on above suggestions.

@CarolEidt @eerhardt @tannergooding

fiigii commented on src/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse41.cs:
```diff
@@ -283,7 +283,7 @@ public static class Sse41
         ///   INSERTPS xmm, xmm/m32, imm8
         /// </summary>
-        public static Vector128<float> Insert(Vector128<float> value, float data, byte index) => Insert(value, data, index);
+        public static Vector128<float> Insert(Vector128<float> value, Vector128<float> data, byte index) => Insert(value, data, index);
```
Corrected the SSE4.1 `Insert` API.

tannergooding replied:
Should we have both overloads or will we specially handle the `Sse.SetScalarVector128` and corresponding `index` masks to not zero the upper bits of `data`?

tannergooding replied:
Also, does the corresponding Avx API need to be updated as well?

fiigii replied:
> Should we have both overloads 

Probably not, SSE4.1 `insertps` has really special semantics that inserts a value selected from the second vector argument into the first vector. But other SSE4.1 "insert" instructions direct insert a scalar value. https://msdn.microsoft.com/en-us/library/bb514071(v=vs.120).aspx

> Also, does the corresponding Avx API need to be updated as well?

No, AVX insert instructions are "normal" :)

fiigii commented:
Addressed the above feedback and fixed SSE4.1 `Insert` API (temporarily disabled its tests).
@CarolEidt @tannergooding @eerhardt Could you please take a look when you get a chance? This change is blocking some HW intrinsic work.

eerhardt commented on src/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse42.cs:
```diff
@@ -18,259 +18,386 @@ public static class Sse42
-        ///   PCMPISTRI xmm, xmm/m128, imm8
-        /// int _mm_cmpistrs (__m128i a, __m128i b, const int imm8)
+        /// </summary>
+        public static bool CompareNotCAndNotZFlag(Vector128<sbyte> left, Vector128<sbyte> right, StringComparisonMode mode) => CompareNotCAndNotZFlag(left, right, mode);
```
In case my original reply gets lost in the `outdated` section, I've copied it here:

I assume `C` == `Carry`, `Z` == `Zero`, `O` == `Overflow`, and `S` == `Sign`?  Can we spell out the words?  Or even better, if we could come up a more descriptive name of what the method does, that would be great.

Reading https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&text=_mm_cmpestr&expand=814,813,814,813, the only difference between `S` and `Z` that I can tell is which argument to check for a null character:

```
Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in a was null, and 0 otherwise. 
```
vs.
```
Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise. 
```

It seems the name should indicate/describe the differences, instead of using `SFlag` or `ZFlag`, and then forcing the caller to look up what `SFlag` means.

fiigii replied:
> I assume C == Carry, Z == Zero, O == Overflow, and S == Sign? Can we spell out the words?

@eerhardt Right, and I am OK to spell them out if it matches the C# naming convention better.

> Or even better, if we could come up a more descriptive name of what the method does, that would be great.

I do not think so. "A more descriptive name" may be too long and not clear. For example, `NotCAndNotZFlag` looks even clearer than "absolute value of lb is larger than or equal to MaxSize and the resulting mask is equal to zero". Meanwhile, we already assume that the users of HW intrinsics have been familar/mastering with the C++ counterparts or ISA design. So, I still suggest to name these intrinsics with flags.

eerhardt replied:
I still think we can come up with better names.  It appears there are 5 cases we need to name:

| Flag   |      Description      |  Possible Names |
|----------|:-------------:|------:|
| S <br>&<br> Z |  Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if any character in a was null, and 0 otherwise. <br>&<br> Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise.  | CompareLeftContainsNull, CompareReturnLeftContainsNull, CompareReturnPartialLeft, CompareGetIsPartialLeft, CompareIsLeftPartial <br>(Same for `Right`) |
| O | Compare packed strings with implicit lengths in a and b using the control in imm8, and returns bit 0 of the resulting bit mask.  |    CompareReturnFirstBit, CompareGetFirstBit |
| C | Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if the resulting mask was non-zero, and 0 otherwise.  |    CompareIsResultNonZero, CompareHasNonZeroResult, CompareNonZeroMask, CompareHasResult  |
| A | Compare packed strings with implicit lengths in a and b using the control in imm8, and returns 1 if b did not contain a null character and the resulting mask was zero, and 0 otherwise.   |      |

`A` is definitely hard to describe succinctly, since it is the combination of two other ones.  So coming up with a good name for it will probably be based on the two underlying names.

-------
**Note**: The [IntrinsicsGuide's description](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&expand=883,2543,936,814&techs=SSE4_2&text=_mm_cmp) for the explicit length `S` & `Z` overloads (`_mm_cmpestrs` & `_mm_cmpestrz`) appears to be incorrect.
`Compare packed strings in a and b with lengths la and lb using the control in imm8, and returns 1 if any character in b was null, and 0 otherwise. `
But then the Operation says:
```
size := (imm8[0] ? 16 : 8) // 8 or 16-bit characters
UpperBound := (128 / size) - 1

dst := (lb <= UpperBound)
```
Where the Operation appears to match what is in the underlying instruction manuals.

-------

In light of the implicit vs explicit differences, to name `S` & `Z`, that's why I like something like `Is Partial` because it describes at a higher-level what is being checked.  ex. `File.Exists` instead of `File.StatReturnsZero`.

fiigii replied:
@eerhardt Thank you so much for the investigation. I read the manual again and I agree that "a more descriptive name" is very necessary because FLAGS is used in a non-standard manner with these SSE4.2 string instructions.
According to the above table you provided, there is a potential name matrix:

| original names | new names |
| -----------------|--------------|
| CompareCFlag  | CompareNonZeroResultMask |
| CompareZFlag | CompareRightContainsNull |
| CompareSFlag | CompareLeftContainsNull |
| CompareOFlag | CompareReturnFirstResultBit |
| CompareNotCAndNotZFlag | CompareZeroResultMaskAndRightNotContainsNull|

Does it look good to you?

cc @CarolEidt @tannergooding

eerhardt replied:
The only issue I see if the difference between implicit and explicit length overloads. If we called the explicit length overloads `ContainsNull` that would be incorrect. It instead returns if the `rightLength` is less than the maximum (8 or 16). So we should use a different name for the explicit length overloads.

I'd like to keep them using the same name, if a decent name comes up. But if one isn't available, I'm fine with the implicit and explicit length overloads being named different (`RightContainsNull` vs. `RightLessThanMax`/`RightIsPartial`/etc.)

tannergooding replied:
I also don't like `ZeroResultMask` as the name here... `CompareEqual` or `IsMatch` might be better?

eerhardt replied:
How about `HasMatch` or `ContainsMatch`? Since depending on the string comparison mode, the whole thing may not match.

fiigii replied:
> I also don't like ZeroResultMask as the name here... CompareEqual or IsMatch might be better?

Zero result masks do not always mean equal or not, which may be controlled by `StringComparisonMode` (e.g., `EqualAny`, `Ranges`, `XXXPolarity`, etc).

eerhardt replied:
`HasResult`/`ContainsResult`?

tannergooding replied:
I think we could make it more obvious by calling the `Polarity` flags something like `NegateResult` instead.

Reading `if (ContainsMatch(op1, op2, StringComparisonMode.NegateResult))` seems much better (both readable and understandable) than `if (CompareZeroResultMask(op1, op2, StringComparisonMode.NegatiePolarity))`

fiigii replied:
> We could do EndOfString, since that applies to both Explicit (where end of string is length determined) and Implicit (where end of string is determined by the null character)?

I think this is a good idea and I also perfer to use the same name for Explicit and Implicit.
How about `Left/RightTerminated`?

| original names | new names |
| -----------------|--------------|
| CompareCFlag  | CompareNonZeroResultMask |
| CompareZFlag | CompareRightTerminated |
| CompareSFlag | CompareLeftTerminated |
| CompareOFlag | CompareReturnFirstResultBit |
| CompareNotCAndNotZFlag | CompareZeroResultMaskAndRightNotTerminated|

fiigii replied:
> I think we could make it more obvious by calling the Polarity flags something like NegateResult instead.

Good point! We can rename these two polarity flags both:
```csharp
        /// <summary>
        /// _SIDD_NEGATIVE_POLARITY
        /// </summary>
        NegativeResult = 0x10,

        /// <summary>
        /// _SIDD_MASKED_NEGATIVE_POLARITY
        /// </summary>
        NegativeUsefulResult = 0x30,
```

fiigii replied:
With the above renamed StringComparisonMode, we seems to have the new name matrix:

| original names | new names |
| -----------------|--------------|
| CompareCFlag  | CompareHasMatch |
| CompareZFlag | CompareRightTerminated |
| CompareSFlag | CompareLeftTerminated |
| CompareOFlag | CompareReturnFirstResultBit |
| CompareNotCAndNotZFlag | CompareNoMatchAndRightNotTerminated|

tannergooding commented:
It might be good to split this into two PRs. One fixing the argument types and the other fixing the SSE4.2 API names.

Did we also already have the discussion on whether it was better for mask/control parameters to be the same as the base type or if it was better to be the same-sized integer type? CC. @eerhardt for input on the API side of things.

fiigii commented:
> It might be good to split this into two PRs. 

I am not sure. I put them together just for simplifying the CoreFX/CoreCLR synchronization.

tannergooding commented:
>  I put them together just for simplifying the CoreFX/CoreCLR synchronization.

Right, but it sounds like the Sse4.2 changes are a bit more complex, and should probably have a bit more "in-depth" review.

I'm fine with keeping them together, if you prefer, but it might be easier to do two separate changes.

fiigii commented:
> but it sounds like the Sse4.2 changes are a bit more complex, and should probably have a bit more "in-depth" review.

@tannergooding Thanks for clarifying. I agree that Sse4.2 changes are a bit more complex and it may impact other API naming. For example, if we decide not using "flags" (e.g., `C` or `Carry`, `Z` or `Zero`, etc.) as @eerhardt suggested, `TestC` and `TestZ` probably also need to be changed to match this convention.

I would split these changes if Sse41 blocks other work.

eerhardt commented on src/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Sse42.PlatformNotSupported.cs:
```diff
@@ -18,259 +18,386 @@ public static class Sse42
         ///   PCMPISTRM xmm, xmm/m128, imm8
         /// </summary>
-        public static Vector128<ushort> CompareImplicitLengthBitMask(Vector128<sbyte> left, Vector128<sbyte> right, StringComparisonMode mode) { throw new PlatformNotSupportedException(); }
+        public static Vector128<ushort> CompareBitMask(Vector128<sbyte> left, Vector128<sbyte> right, StringComparisonMode mode) { throw new PlatformNotSupportedException(); }
```
For my information - Why did we break out `BitMask` vs `UnitMask` into different overloads?  Why not keep it in the `StringComparisonMode` enum?

fiigii replied:
Ah, good catch! This is a legacy design from the previous versions, will fix. Thanks!

CarolEidt commented on src/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Enums.cs:
```diff
@@ -167,23 +168,56 @@ public enum FloatComparisonMode : byte
+        /// <summary>
+        /// _SIDD_UNIT_MASK
+        /// </summary>
+        UnitMask = 0x40,
```
How are `UnitMask` and `MostSignificant` (or `LeastSignificant` and `BitMask`) distinguished? I don't think we should have multiple names for the same value in a single enum - are these used in different contexts? If so, I would think they'd be different enums.

fiigii replied:
> are these used in different contexts?

Yes.
`UnitMask` and `BitMask` is only used for `CompareMask` (`PCMPESTRM`).
`LeastSignificant` and `MostSignificant` is only used for `CompareIndex` (`PCMPESTRI`).

fiigii replied:
> If so, I would think they'd be different enums.

So, `CompareMask` and `CompareIndex` would have an additional imm parameter.
How about encoding these two flags into the intrinsic functiom name?

CarolEidt replied:
> How about encoding these two flags into the intrinsic function name?

I don't think that's necessary - I just don't see why the enums should be the same.

eerhardt replied:
I think it depends on whether we want to map closely with the underlying hardware instruction, which I thought was a desire.

Section `4.1.5 Output Selection` in https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-2b-manual.pdf

![image](https://user-images.githubusercontent.com/8291187/41362551-451ca21a-6ef7-11e8-9001-23e5c0f77d72.png)

Since it will be a `Flags` enum (after this PR), I don't think this is a huge issue. It follows the underlying HW instruction design, and is understandable - use Least|Most Significant with the `CompareIndex` method, use Bit|Unit Mask with the `CompareMask` method.

tannergooding replied:
@eerhardt, I think @CarolEidt's point is that it is still consistent with the underlying instructions. But is exposing it to the user in a more "safe" and "understandable" manner.

You can certainly do `Bit|Unit`  with CompareIndex (instead of `Least|Most`) and it will do the "correct" thing, but it would make the code harder to understand/etc.

fiigii replied:
> You can certainly do Bit|Unit with CompareIndex (instead of Least|Most) and it will do the "correct" thing, but it would make the code harder to understand/etc.

Yes, but there is a trade-off between safer/stronger enum type-checking and usability. I think it may not worthwhile to complicate the APIs just for these two flags.

tannergooding replied:
@fiigii, is it really that more complex?

On the API side, we'll have two enums, that are restricted to the Index vs Mask overloads (which we already have).

On the JIT side, we can just ensure that we carry it around as a byte constant and don't need to worry about there being two enums.

fiigii replied:
> On the API side, we'll have two enums,
> On the JIT side, we can just ensure that we carry it around as a byte constant

Don't you mean define `LeastSignificant`/`BitMask` as two enum types? 

```csharp
    public enum StringComparisonMode : byte {
        // ...
    }

    public enum IndexMode : byte {
        // ...
        /// <summary>
        /// _SIDD_LEAST_SIGNIFICANT
        /// </summary>
        LeastSignificant = 0x00,

        /// <summary>
        /// _SIDD_MOST_SIGNIFICANT
        /// </summary>
        MostSignificant = 0x40,
    }

    public enum MaskMode : byte {
        // ...
        /// <summary>
        /// _SIDD_BIT_MASK
        /// </summary>
        BitMask = 0x00,

        /// <summary>
        /// _SIDD_UNIT_MASK
        /// </summary>
        UnitMask = 0x40,
    }
```

If so, these APIs would need an additional imm-parameter, right?

tannergooding replied:
> If so, these APIs would need an additional imm-parameter, right?

Right.

We might also be able to just double the number of exposed methods and encode the "mode" in the name, if @eerhardt and @CarolEidt thought that was more reasonable.

eerhardt replied:
>> I think @CarolEidt's point is that it is still consistent with the underlying instructions

>If so, these APIs would need an additional imm-parameter

> We might also be able to just double the number of exposed methods and encode the "mode" in the name

At what point does it become "inconsistent" with the underlying instructions?

tannergooding replied:
> At what point does it become "inconsistent" with the underlying instructions?

IMO, it would be if we no longer guaranteed the particular instruction would be emitted or when we no longer made it possible to cause certain functionality to happen (like if we excluded one of the possible flag values).

tannergooding replied:
I think that exposing something like `GetIndexOfFirstMatch` is completely within the realm of matching the underlying instruction and it mirrors other intrinsic APIs where we are doing similar things (like `Sse.Compare*OrderedScalar`, where all 6+ overloads are exposed, rather than providing an enum that the user passes in for the Comparisonmode).

fiigii replied:
Finally, these two solutions complicate API design (@eerhardt 's solution) or JIT implementation (@tannergooding's solution, especially for non-const fallback). So, I said that may not worthwhile for such bit type safety.

eerhardt replied:
How about we do the following?  @tannergooding or @CarolEidt, if you want the enum to change, can you open a new corefx issue with an API proposal for changing the enum? That will probably be the best way forward. Designing it in this PR probably isn't the best solution.

tannergooding replied:
> Using an enum allows me to use the API without making the choice.

I'm not sure how valid this is. The choice can make a fairly big difference on how you write your algorithm, so you will generally need to make the choice and be aware of how it impacts your code, right out.

(You will also have to specify one of the two enum values either way)

eerhardt replied:
> I'm not sure how valid this is.

It's completely valid when I'm exploring the API to learn how it works.

> (You will also have to specify one of the two enum values either way)

Only in the 3 enums design.  As currently proposed I wouldn't need to specify `Bit/UnitMask` when calling `CompareMask`.

tannergooding replied:
> As currently proposed I wouldn't need to specify Bit/UnitMask when calling CompareMask.

I'm not convinced that's a good thing 😄

tannergooding replied:
That is, these APIs and instructions are already fairly hard to understand. Making parts of it more obvious, even if it results in us exposing double the methods might be a good thing (It looks like we have ~50 methods exposed right now).

CarolEidt commented:
At this point, I would propose changing this PR so that it *doesn't* change the enum names or usage (i.e. only do changes 1 and 3.
There should be an API discussion in corefx before settling in the names and usage of the various result and comparison modes.

fiigii commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Sse41.cs:
```diff
@@ -182,12 +182,12 @@ public static class Sse41
         ///   PEXTRB reg/m8, xmm, imm8
         /// </summary>
-        public static sbyte Extract(Vector128<sbyte> value, byte index) => Extract(value, index);
+        public static int Extract(Vector128<sbyte> value, byte index) => Extract(value, index);
```
Also fixes SSE2/SSE4.1/AVX `Extract` return type to reflect the underlying instruction behavior https://github.com/dotnet/coreclr/issues/17957

cc @CarolEidt @tannergooding

tannergooding commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx.cs:
```diff
@@ -239,59 +239,59 @@ public static class Avx
                 sbyte* buffer = stackalloc sbyte[32];
                 Store(buffer, value);
-                return buffer[index];
+                return (int)buffer[index];
```
Question for future: Is this more efficient than a GetUpper and Extract combo?

fiigii replied:
No, it's not. Here managed implementation is just the non-const fallback that simplifies JIT implementation of the imm-helpers.

jkotas replied:
Can this be just `return ((sbyte*)(&value))[index];` ?

tannergooding replied:
No, you can't take the address of a generic.

tannergooding replied:
We could probably do a one line with `S.R.CS.Unsafe`

jkotas replied:
`return Unsafe.Add<ref Unsafe.As<Vector256<sbyte>, sbyte>(ref value), index & 0x1F);`

tannergooding commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Sse41.PlatformNotSupported.cs:
```diff
@@ -182,12 +182,12 @@ public static class Sse41
         ///   PEXTRB reg/m8, xmm, imm8
         /// </summary>
-        public static sbyte Extract(Vector128<sbyte> value, byte index) { throw new PlatformNotSupportedException(); }
+        public static int Extract(Vector128<sbyte> value, byte index) { throw new PlatformNotSupportedException(); }
```
What impact does this have on containment, since it extracts to `reg/m8` and zero extends the upper bits...

fiigii replied:
Not sure. There are some other similar instructions (e.g., PMOVZX* and PMOVSX* ), I will investigate later.

jkotas replied:
What's the advantage of returning `int` instead of `sbyte`? 

The `int` is going to be zero extended byte that is pretty non-intuitive.

tannergooding replied:
@jkotas, the behavior of the `PEXTRB` instruction is to zero extend:
> Extract a byte integer value from xmm2 at the source byte offset specified by imm8 into reg or m8. The upper bits of r32 or r64 are zeroed.`

jkotas replied:
Also, does the manual implementation zero extends the same way? I think the current manual implementation going to sign-extend, not zero-extended.

jkotas replied:
> the behavior of the PEXTRB instruction is to zero extend

Right. And it makes the behavior very non-intuitive when it returns int. I think this instruction should be either defined for `Vector128<byte>` only; or it should return `sbyte`.

mikedn replied:
The manual implementation will probably sign-extend exactly because the return type was `sbyte` instead of `int`. Thing is, the instruction extracts a byte into a 32 bit register, it doesn't exactly deal with signedness. To get `sbyte` to work properly you'd need to add another instruction - `movsx eax, al` - that's a bit unusual for an intrinsic.

tannergooding replied:
@jkotas, it was being changed from `sbyte` to `int` due to `https://github.com/dotnet/coreclr/issues/17957`.

However, I agree that just exposing `byte` might make more sense. It might also reduce confusion for the case where we write directly to memory, and zero extension cannot happen.

fiigii replied:
@jkotas Thanks for capturing this. 

I also agree to just remove the `sbyte` and `short` overloads rather than changing the return type to `int/uint`.

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Avx/Avx_ro.csproj:
```diff
@@ -44,14 +44,6 @@
     <Compile Include="DuplicateEvenIndexed.Double.cs" />
     <Compile Include="DuplicateEvenIndexed.Single.cs" />
     <Compile Include="DuplicateOddIndexed.Single.cs" />
-    <Compile Include="Extract.Byte.1.cs" />
```
I wonder  if just commenting these out would be better, since we will need to add them back once the changes flow to CoreFX and back again....

fiigii replied:
Ah, yes, but I already removed them 😄

tannergooding replied:
Might be good to fix it, so we don't have +10k/-10k diffs for the same code in such a brief period...

fiigii commented on src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Avx.cs:
```diff
@@ -235,64 +235,23 @@ public static class Avx
-                Store(buffer, value);
-                return buffer[index];   
-            }
+            return Unsafe.Add<byte>(ref Unsafe.As<Vector256<byte>, byte>(ref value), index & 0x1F);
```
Removed `sbyte` and `short` overloads of SSE2/SSE4.1/AVX `Extract` and simplify `Avx.Extract` non-const fallback as @jkotas's suggestion.

I would prepare the CoreFX counterpart if this PR looks good to you guys.

fiigii commented:
@eerhardt Can we merge this PR?

eerhardt commented:
Oops - I thought this was already merged.  Merging now.
